server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java

```diff
@@ -0,0 +1,315 @@
+
+  public static final String MSG_RECONNECTION_FAIL =
+      "Fail to connect to any config node. Please check server it";
+
+  private ConfigIService.Iface client;
+
+  private TTransport transport;
+
+  private Endpoint configLeader;
+
+  private List<Endpoint> configNodes;
+
+  public ConfigNodeClient() throws BadNodeUrlException, IoTDBConnectionException {
+    // Read config nodes from configuration
+    configNodes =
+        CommonUtils.parseNodeUrls(IoTDBDescriptor.getInstance().getConfig().getConfigNodeUrls());
+    init();
+  }
+
+  public ConfigNodeClient(List<Endpoint> configNodes) throws IoTDBConnectionException {
+    this.configNodes = configNodes;
+    init();
+  }
+
+  public ConfigNodeClient(List<Endpoint> configNodes, Endpoint configLeader)
+      throws IoTDBConnectionException {
+    this.configNodes = configNodes;
+    this.configLeader = configLeader;
+    init();
+  }
+
+  public void init() throws IoTDBConnectionException {
+    reconnect();
+  }
+
+  public void connect(Endpoint endpoint) throws IoTDBConnectionException {
+    try {
+      transport =
+          RpcTransportFactory.INSTANCE.getTransport(
+              // as there is a try-catch already, we do not need to use TSocket.wrap
+              endpoint.getIp(), endpoint.getPort(), TIMEOUT_MS);
+      transport.open();
+    } catch (TTransportException e) {
+      throw new IoTDBConnectionException(e);
+    }
+
+    if (IoTDBDescriptor.getInstance().getConfig().isRpcThriftCompressionEnable()) {
+      client = new ConfigIService.Client(new TCompactProtocol(transport));
+    } else {
+      client = new ConfigIService.Client(new TBinaryProtocol(transport));
+    }
+  }
+
+  private void reconnect() throws IoTDBConnectionException {
+    if (configLeader != null) {
+      try {
+        connect(configLeader);
+        return;
+      } catch (IoTDBConnectionException e) {
+        logger.error("The current node may have been down {},try next node", configLeader);
+        configLeader = null;
+      }
+    }
+
+    Random random = new Random();
+    if (transport != null) {
+      transport.close();
+    }
+    int currHostIndex = random.nextInt(configNodes.size());
+    int tryHostNum = 0;
+    for (int j = currHostIndex; j < configNodes.size(); j++) {
+      if (tryHostNum == configNodes.size()) {
+        break;
+      }
+      Endpoint tryEndpoint = configNodes.get(j);
+      if (j == configNodes.size() - 1) {
+        j = -1;
+      }
+      tryHostNum++;
+      try {
+        connect(tryEndpoint);
+        return;
+      } catch (IoTDBConnectionException e) {
+        logger.error("The current node may have been down {},try next node", tryEndpoint);
+      }
+    }
+    throw new IoTDBConnectionException(MSG_RECONNECTION_FAIL);
+  }
+
+  public void close() {
+    transport.close();
+  }
+
+  private boolean verifyNeedRedirect(TSStatus status) throws StatementExecutionException {
+    if (status.getCode() == TSStatusCode.NEED_REDIRECTION.getStatusCode()) {
```

Comment:
I set TSStatus.message when its TSStatusCode is NEED_REDIRECTION,  which you can use in logger